code/modules/vehicles/tonk.dm

```diff
@@ -0,0 +1,354 @@
+/*
+TONKS BY KMC
+THIS IS LIKE REGULAR CM TONK BUT IT'S LESS SHIT
+WHOEVER MADE CM TANKS: YOU ARE A BAD CODER!!!!!
+*/
+
+//generic component stuff
+/atom
+	var/tank_crash_damage = 10  //Arbitrary number. Set this to whatever you want with extra walltypes
+
+/atom/proc/tank_act(var/obj/vehicle/tonk/victim)
+	visible_message("[victim] CRASHES INTO [src]")
+	victim.health -= tank_crash_damage
+	qdel(src)
+
+/obj/item/weapon/tank_weapon //These are really just proof of concept weapons. You'll probably want to adapt them to use the hardpoint crap that CM has.
+	name = "TGS 4 main tank cannon"
+	desc = "A gun that works about 50% of the time, but at least it's open source! It fires tank shells."
+	icon = 'icons/obj/hardpoint_modules.dmi'
+	icon_state = "glauncher"
+	var/obj/item/ammo_magazine/ammo = new /obj/item/ammo_magazine/tank/ltb_cannon
+	var/obj/vehicle/tonk/owner
+	var/list/fire_sounds = list('sound/weapons/tank_cannon_fire1.ogg', 'sound/weapons/tank_cannon_fire2.ogg')
+	var/fire_delay
+	var/cooldown = 60 //6 second weapons cooldown
+	var/lastfired = 0 //When were we last shot?
+
+/obj/item/weapon/tank_weapon/minigun
+	name = "TGS 3 XENOCRUSHER tank machine gun"
+	desc = "A much better gun that shits out bullets at ridiculous speeds, don't get in its way!"
+	icon = 'icons/obj/hardpoint_modules.dmi'
+	icon_state = "m56_cupola"
+	ammo = new /obj/item/ammo_magazine/tank/ltaaap_minigun
+	fire_sounds = list('sound/weapons/tank_minigun_loop.ogg')
+	cooldown = 2 //Minimal cooldown
+
+/obj/item/weapon/tank_weapon/Initialize()
+	. = ..()
+
+/obj/item/weapon/tank_weapon/proc/fire(var/atom/T,var/mob/user)
+	if(!can_fire(T))
+		return
+	if(ammo.current_rounds <= 0)
+		return FALSE
+	if(world.time < lastfired + cooldown)
+		if(owner.gunner)
+			return FALSE
+		return FALSE
+	lastfired = world.time
+	var/obj/item/projectile/P = new
+	P.generate_bullet(new ammo.default_ammo)
+	log_combat(user, user, "fired the [src].")
+	log_explosion("[user] fired the [src] at [AREACOORD(loc)].")
+	P.fire_at(T, owner, src, P.ammo.max_range, P.ammo.shell_speed)
+	if(!CONFIG_GET(flag/tank_mouth_noise))
+		playsound(get_turf(owner), pick(fire_sounds), 60, 1)
+	ammo.current_rounds--
+	return TRUE
+
+/obj/item/weapon/tank_weapon/proc/can_fire(var/turf/T)
+	if(T in orange(owner, 2))
+		if(owner.gunner)
+			to_chat(owner.gunner, "Firing your main gun here could damage the tank!")
+			return FALSE
+	return TRUE
```

Comment:
Wouldn't it make sense to check for the existence of `owner.gunner` here instead of  in `fire()`?